notebooks/machine_learning/raw/tut8.ipynb

```diff
@@ -0,0 +1,479 @@
+   "metadata": {},
+   "source": [
+    "## Confusion Matrix\n",
+    "\n",
+    "Our model did pretty well, correctly predicting around 86% of the price ranges in the validation data. It's often useful to look at where the model is failing with a **confusion matrix** which shows us how our model classified the inputs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Confusion matrix:\n",
+      "[[130   6   0   0]\n",
+      " [  4  91  15   0]\n",
+      " [  0  21  98  12]\n",
+      " [  0   0  10 113]]\n",
+      "\n",
+      "Normalized confusion matrix:\n",
+      "[0.95588235 0.04411765 0.         0.        ]\n",
+      "[0.03636364 0.82727273 0.13636364 0.        ]\n",
+      "[0.         0.16030534 0.7480916  0.09160305]\n",
+      "[0.         0.         0.08130081 0.91869919]\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Calculate the confusion matrix itself\n",
+    "confusion = metrics.confusion_matrix(val_y, pred_y)\n",
+    "print(f\"Confusion matrix:\\n{confusion}\")\n",
+    "\n",
+    "\n",
+    "# Normalizing by the true label counts to get rates\n",
+    "print(f\"\\nNormalized confusion matrix:\")\n",
+    "for row in confusion:\n",
+    "    print(row / row.sum())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<img src=\"https://i.imgur.com/idD0k8y.png\" alt=\"example confusion matrix\" width=400px>\n",
+    "\n",
+    "The rows of the confusion matrix are the true class and the columns are the predicted class. The diagonal tells us how many of each class the model predicted correctly. The off-diagonals show where the model is making wrong predictions, where it is \"confused.\" For example, looking at the first column and second row, we classified four phones that were actually low cost as medium cost. We see for classes 0 and 3, the low cost and highest cost phones, our model works really well, above 90% accurate. However, our model is weaker for medium and high cost phones. Note that incorrect predictions are only between adjacent classes. The model doesn't confuse low cost and very high cost phones.\n",
+    "\n",
+    "The diagonal of the normalized confusion matrix gives us the **true positive rate** for each of the classes. This is the fraction of positive cases in the validation data the model predicted to be positive. This value is important because the accuracy metric can sometimes hide information you care about. Occasionally you'll have unbalanced classes where you have many more examples of one class compared to another. For example, it's typical that when you're trying to diagnose a disease with a machine learning model, most examples in your data won't have that disease. Your model could be very good at identifying when people don't have the disease (negative cases), but poor at identifying positive cases. \n",
```

Comment:
I'm skeptical of this characterization that things on the diagonal are the "True Positive Rate" in a multi-class setting.

As you say in the next sentence, the "true positive rate" is the fraction of actual positives that are predicted to be positive. But some parts of the diagonal are neither predicted nor actually "positive." 

In fact, in the multi-class case, the characterization of anything as a truly the "positive" class seems awkward.

Anyway, the diagonal could be called the "correctly classified" but not "True positive."

I think all of that is unnecessary though.